src/explicit/input-configurations.hh

```diff
@@ -0,0 +1,304 @@
+              throw std::logic_error("in explicit_::ConvexShapeContact: object "
+                                     "contact surfaces should be hold by the "
+                                     "same object");
+            }
+          }
+          for (JointAndShapes_t::const_iterator it(floorSurfaces.begin());
+               it != floorSurfaces.end(); ++it)
+          {
+            JointPtr_t joint1(it->first);
+            std::vector<bool> conf (robot->configSize(), false);
+            while (joint1 && joint1->index() != 0) {
+              for (size_type i = 0; i < joint1->configSize(); ++i)
+                conf[joint1->rankInConfiguration() + i] =
+                  !conf[joint1->rankInConfiguration() + i];
+              hppDout (info, "Adding joint1 " << joint1->name ()
+                       << " as input variable to ConvexShapeContact explicit "
+                       "constraint.");
+              joint1 = joint1->parentJoint();
+            }
+            JointPtr_t joint2(objectJoint->parentJoint());
+            while (joint2 && joint2->index() != 0) {
+              for (size_type i = 0; i < joint2->configSize(); ++i)
+                conf[joint2->rankInConfiguration() + i] =
+                  !conf[joint2->rankInConfiguration() + i];
+              hppDout (info, "Adding joint2 " << joint2->name ()
+                       << " as input variable to ConvexShapeContact explicit "
+                       "constraint.");
+              joint2 = joint2->parentJoint();
+            }
+            for (std::size_t i=0; i<inputConf.size(); ++i)
+            {
+              if (conf[i]) inputConf[i] = true;
+            }
+          }
+          size_type res(0);
+          for (std::size_t i=0; i<inputConf.size(); ++i)
+          {
+            if (inputConf[i]) ++res;
+          }
+          return res;
+        }
+        // Compute input configuration variables of
+        // explicit_::ConvexShapeContact.
+        inline BlockIndex::segments_t inputVelocityVariables
+        (const DevicePtr_t& robot, const JointAndShapes_t& floorSurfaces,
+         const JointAndShapes_t& objectSurfaces)
+        {
+          std::vector<bool> res (robot->numberDof(), false);
+          assert(objectSurfaces.size()>0);
+          JointPtr_t objectJoint = objectSurfaces.front().first;
+          for (JointAndShapes_t::const_iterator it(floorSurfaces.begin());
+               it != floorSurfaces.end(); ++it)
+          {
+            JointPtr_t joint1(it->first);
+            std::vector<bool> conf (robot->numberDof(), false);
+            while (joint1 && joint1->index() != 0) {
+              for (size_type i = 0; i < joint1->numberDof(); ++i)
+                conf[joint1->rankInVelocity() + i] =
+                  !conf[joint1->rankInVelocity() + i];
+              hppDout (info, "Adding joint1 " << joint1->name ()
+                       << " as input variable.");
+              joint1 = joint1->parentJoint();
+            }
```

Comment:
Overall, in this file, there is a lot of code which *looks* duplicated. For instance, this loop is repeated plenty of times in this file. It may be a good idea to put it in a function. It would help having a more readable code.